In [ ]:
!pip install mlxtend

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Task 1: Data Preprocessing

In [ ]:
df = pd.read_csv('/content/Copy of Online retail10.csv')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                             burgers,meatballs,eggs                                                                                                                                                                             
1                                            chutney                                                                                                                                                                             
2                                     turkey,avocado                                                                                                                                                                             
3  mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                     low fat yogurt

In [ ]:
df.describe()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
count                                                7500                                                                                                                                                                             
unique                                               5175                                                                                                                                                                             
top                                               cookies                                                                                                                                                                             
freq                                                  223

In [ ]:
# Handling missing values
df = df.dropna()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
df=pd.get_dummies(df)
df.head()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil_almonds  \
0                                              False                                                                                                                                                                                         
1                                              False                                                                                                                                                                                         
2                                              False                                                                                                                                                                                         
3                                              False                                                                                                                                                                                         
4                                              False                                                                                                                                                                                         

   shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil_almonds,cake,low fat yogurt  \
0                                              False                                                                                                                                                                                                             
1                                              False                                                                                                                                                                                                             
2                                              False                                                                                                                                                                                                             
3                                              False                                                                                                                                                                                                             
4                                              False                                                                                                                                                                                                             

   shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil_almonds,cookies  \
0                                              False                                                                                                                                                                                                 
1                                              False                                                                                                                                                                                                 
2                                              False                                                                                                                                                                                                 
3                                        

Task 2: Association Rule Mining

In [ ]:
transactions = [
    ['milk', 'bread', 'eggs'],
    ['milk', 'bread'],
    ['milk', 'eggs', 'butter'],
    ['bread', 'butter']
]

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_transformed = pd.DataFrame(te_ary, columns=te.columns_)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Apply the Apriori algorithm
min_support = 0.01
frequent_itemsets = apriori(df_transformed, min_support=min_support, use_colnames=True)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Generate association rules
min_confidence = 0.2
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)
print(rules.head())

  antecedents consequents  antecedent support  consequent support  support  \
0    (butter)     (bread)                0.50                0.75     0.25   
1     (bread)    (butter)                0.75                0.50     0.25   
2     (bread)      (eggs)                0.75                0.50     0.25   
3      (eggs)     (bread)                0.50                0.75     0.25   
4     (bread)      (milk)                0.75                0.75     0.50   

   confidence      lift  leverage  conviction  zhangs_metric  
0    0.500000  0.666667   -0.1250        0.50      -0.500000  
1    0.333333  0.666667   -0.1250        0.75      -0.666667  
2    0.333333  0.666667   -0.1250        0.75      -0.666667  
3    0.500000  0.666667   -0.1250        0.50      -0.500000  
4    0.666667  0.888889   -0.0625        0.75      -0.333333  


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Task 3: Analysis and Interpretation



antecedents: milk

consequents: bread

support: support of {milk, bread})

confidence: confidence of {milk} -> {bread})

lift: The ratio of observed support to expected support if antecedent and consequent were independent, lift of {milk} -> {bread}

In [ ]:
rules.sort_values('confidence',ascending = False) # can sort as lift value

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/pandas/core/nanops.py:1010: RuntimeWarning: invalid value encountered in subtract
  sqr = _ensure_numeric((avg - values) ** 2)


antecedents     consequents  antecedent support  consequent support  \
13   (bread, eggs)          (milk)                0.25                0.75   
19  (butter, eggs)          (milk)                0.25                0.75   
18  (butter, milk)          (eggs)                0.25                0.50   
11          (eggs)          (milk)                0.50                0.75   
4          (bread)          (milk)                0.75                0.75   
5           (milk)         (bread)                0.75                0.75   
10          (milk)          (eggs)                0.75                0.50   
0         (butter)         (bread)                0.50                0.75   
21        (butter)    (milk, eggs)                0.50                0.50   
20    (milk, eggs)        (butter)                0.50                0.50   
17          (eggs)   (bread, milk)                0.50                0.50   
14    (milk, eggs)         (bread)                0.50                0.75   
12   (bread, milk)          (eggs)                0.50                0.50   
8         (butter)          (milk)                0.50                0.75   
7           (eggs)        (butter)                0.50                0.50   
6         (butter)          (eggs)                0.50                0.50   
3           (eggs)         (bread)                0.50                0.75   
23          (eggs)  (butter, milk)                0.50                0.25   
1          (bread)        (butter)                0.75                0.50   
9           (milk)        (butter)                0.75                0.50   
15         (bread)    (milk, eggs)                0.75                0.50   
16          (milk)   (bread, eggs)                0.75                0.25   
2          (bread)          (eggs)                0.75                0.50   
22          (milk)  (butter, eggs)                0.75                0.25   

    support  confidence      lift  leverage  conviction  zhangs_metric  
13     0.25    1.000000  1.333333    0.0625         inf       0.333333  
19     0.25    1.000000  1.333333    0.0625         inf       0.333333  
18     0.25    1.000000  2.000000    0.1250         inf       0.666667  
11     0.50    1.000000  1.333333    0.1250         inf       0.500000  
4      0.50    0.666667  0.888889   -0.0625       0.750      -0.333333  
5      0.50    0.666667  0.888889   -0.0625       0.750      -0.333333  
10     0.50    0.666667  1.333333    0.1250       1.500       1.000000  
0      0.25    0.500000  0.666667   -0.1250       0.500      -0.500000  
21     0.25    0.500000  1.000000    0.0000       1.000       0.000000  
20     0.25    0.500000  1.000000    0.0000       1.000       0.000000  
17     0.25    0.500000  1.000000    0.0000       1.000       0.000000  
14     0.25    0.500000  0.666667   -0.1250       0.500      -0.500000  
12     0.25    0.500000  1.000000    0.0000       1.000       0.000000  
8      0.25    0.500000  0.666667   -0.1250       0.500      -0.500000  
7      0.25    0.500000  1.000000    0.0000       1.000       0.000000  
6      0.25    0.500000  1.000000    0.0000       1.000       0.000000  
3      0.25    0.500000  0.666667   -0.1250       0.500      -0.500000  
23     0.25    0.500000  2.000000    0.1250       1.500       1.000000  
1      0.25    0.333333  0.666667   -0.1250       0.750      -0.666667  
9      0.25    0.333333  0.666667   -0.1250       0.750      -0.666667  
15     0.25    0.333333  0.666667   -0.1250       0.750      -0.666667  
16     0.25    0.333333  1.333333    0.0625       1.125       1.000000  
2      0.25    0.333333  0.666667   -0.1250       0.750      -0.666667  
22     0.25    0.333333  1.333333    0.0625       1.125       1.000000

High confidence in the rule {milk} -> {bread} indicates that customers who buy milk often buy bread too. A lift value greater than 1, such as 1.33, suggests these items are bought together more frequently than by chance. High support for {milk, bread} shows they are commonly purchased together, highlighting a strong relationship between them. This insight can guide product placement and promotions to boost sales.

1.	What is lift and why is it important in Association rules?

Ans

Lift is a measure used in association rule mining to determine how much more likely two items are to be bought together than if they were completely independent of each other. It is a ratio of the observed support for the rule to the expected support if the items were independent.

Importance of Lift:

Evaluates Rule Strength: Lift provides insight into how strong an association rule is by comparing the actual co-occurrence of items to what would be expected by random chance.

Prioritization of Rules: Higher lift values indicate stronger rules, which can help in prioritizing marketing strategies or product placements.

2.	What is support and Confidence. How do you calculate them?

Ans:

1. Support
Support measures how frequently an itemset appears in the dataset. It helps identify how common or popular an itemset is

Support tells us how often an itemset appears in the dataset, providing a measure of its frequency.

2. Confidence
It measures the likelihood that a rule is correct. Confidence indicates how often items in 𝑌 appear in transactions that contain 𝑋.

3. What are some limitations or challenges of Association rules mining?

Ans.

While association rule mining is a powerful technique, it has some limitations and challenges:

Scalability: As the number of items increases, the number of possible itemsets grows exponentially, making it computationally expensive to generate and evaluate all possible rules.

Support and Confidence Thresholds: Setting appropriate thresholds for support and confidence can be challenging. Too high, and you might miss interesting rules; too low, and you might be overwhelmed with trivial rules.

High Dimensionality: In datasets with many items, association rule mining can generate a vast number of rules, making it difficult to identify the most useful ones.

Interpretability: The sheer number of generated rules can lead to information overload, making it difficult for users to interpret and act on the results.

Dependency on Historical Data: Association rules are derived from historical data, which may not always represent future behavior accurately, especially in dynamic markets.

Sparsity: In cases where transactions are sparse (i.e., many unique items but few frequent combinations), it can be difficult to find meaningful associations.